temporal-sdk/src/main/java/io/temporal/internal/worker/AsyncPoller.java

```diff
@@ -0,0 +1,387 @@
+          pollerBalancer.startPoll(asyncTaskPoller.getLabel());
+
+          pollRequest
+              .handle(
+                  (task, e) -> {
+                    pollerBalancer.endPoll(asyncTaskPoller.getLabel());
+                    if (e instanceof CompletionException) {
+                      e = e.getCause();
+                    }
+                    pollerSemaphore.release();
+                    pollScaleReportHandle.report(task, e);
+                    if (e != null) {
+                      uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
+                      pollBackoffThrottler.failure(
+                          (e instanceof StatusRuntimeException)
+                              ? ((StatusRuntimeException) e).getStatus().getCode()
+                              : Status.Code.UNKNOWN);
+                      slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), finalPermit);
+                      return null;
+                    }
+                    if (task != null) {
+                      taskExecutor.process(task);
+                    } else {
+                      slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), finalPermit);
+                    }
+                    pollBackoffThrottler.success();
+                    return null;
+                  })
+              .exceptionally(
+                  throwable -> {
+                    log.error("Error while trying to poll task", throwable);
+                    return null;
+                  });
+        } catch (PollTaskAsyncAbort ab) {
+          abort = true;
+        } catch (Throwable e) {
+          if (e instanceof InterruptedException) {
+            // we restore the flag here, so it can be checked and processed (with exit) in finally.
+            Thread.currentThread().interrupt();
+          }
+          uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
+        } finally {
+          // release the slot if it was acquired, but a poll request was not made
+          if (!pollRequestMade) {
+            if (permit != null) {
+              slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), permit);
+            }
+            if (pollerSemaphoreAcquired) {
+              pollerSemaphore.release();
+            }
+          }
+
+          if (shouldTerminate()) {
+            pollerBalancer.removePoller(asyncTaskPoller.getLabel());
+            abort = true;
+          }
+        }
+      }
+      log.info(
+          "Poll loop is terminated: {} - {}",
+          AsyncPoller.this.getClass().getSimpleName(),
+          asyncTaskPoller.getLabel());
+    }
+  }
+
+  /**
+   * PollQueueBalancer is used to ensure that at least one poller is running for each task. This is
+   * necessary to avoid one poller from consuming all the slots and starving other pollers.
+   */
+  @ThreadSafe
+  class PollQueueBalancer {
+    Map<String, AtomicInteger> taskCounts = new HashMap<>();
```

Comment:
Seems like this could store normal ints in the values since usage is always protected by the lock?

Reply:
Ah yep